controllers/cloud.redhat.com/providers/sizing/sizing.go

```diff
@@ -0,0 +1,151 @@
+//a sizing package will be helpful
+
+//Note on naming:
+//Naming is hard. In the context of this API "sizes" are
+//the t shirt sizes (small, medium, etc) whereas "capacities"
+//are the values k8s uses like Gi, M, m, etc. This distinction is
+//important because most of what we're doing here is
+//converting sizes to capacities, so I enforce the distinction
+//strictly. The method names are long, but more importantly, accurate
+
+import (
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+)
+
+const (
+	//The nature of the beast in this code is we got a lot of magic strings
+	XSMALL string = "x-small"
+	SMALL  string = "small"
+	MEDIUM string = "medium"
+	LARGE  string = "large"
+	XLARGE string = "x-large"
+	//We need to define default sizes because if a ClowdApp doesn't provide
+	//volume or ram/cpu capacities we just get an empty string, so we need
+	//defaults to plug in there
+	DEFAULT_SIZE_VOL     string = XSMALL
+	DEFAULT_SIZE_CPU_RAM string = SMALL
+)
+
+// Public methods
+
+//Get default resource requirement requests and limits
+func GetDefaultResourceRequirements() core.ResourceRequirements {
+	return GetResourceRequirementsForSize(GetDefaultSizeCPURAM())
+}
+
+//Gets the default size for CPU and RAM
+func GetDefaultSizeCPURAM() string {
+	return DEFAULT_SIZE_CPU_RAM
+}
+
+//Gets the default vol size
+func GetDefaultSizeVol() string {
+	return DEFAULT_SIZE_VOL
+}
+
+//Get the default volume capacity
+func GetDefaultVolCapacity() string {
+	return getVolSizeToCapacityMap()[GetDefaultSizeVol()]
+}
+
+//Get resource requirements - request and limits - for a given size
+func GetResourceRequirementsForSize(tShirtSize string) core.ResourceRequirements {
+	requestSize := useDefaultIfEmptySize(tShirtSize, GetDefaultSizeCPURAM())
+	cpu := getCPUSizeToCapacityMap()
+	ram := getRAMSizeToCapacityMap()
+	limitSize := getLimitSizeForRequestSize(requestSize)
+	return core.ResourceRequirements{
+		Limits: core.ResourceList{
+			"memory": resource.MustParse(ram[limitSize]),
+			"cpu":    resource.MustParse(cpu[limitSize]),
+		},
+		Requests: core.ResourceList{
+			"memory": resource.MustParse(ram[requestSize]),
+			"cpu":    resource.MustParse(cpu[requestSize]),
+		},
+	}
+}
+
+//For a givin vol size get the capacity. Providing "" gets the default.
+func GetVolCapacityForSize(size string) string {
+	requestSize := useDefaultIfEmptySize(size, GetDefaultSizeVol())
+	return getVolSizeToCapacityMap()[requestSize]
+}
+
+//Accepts 2 sizes. Returns true if first size is larger than second
+func IsSizeLarger(capacityA string, capacityB string) bool {
+	capacities := map[string]int{
+		XSMALL: 0,
+		SMALL:  1,
+		MEDIUM: 2,
+		LARGE:  3,
+		XLARGE: 4,
+	}
+	return capacities[capacityA] > capacities[capacityB]
+}
+
+// Private methods
+
+//Get a map of CPU T-Shirt sizes to capacities
+func getCPUSizeToCapacityMap() map[string]string {
+	return map[string]string{
```

Comment:
Again here, there is no real reason to return these as functions, they could be const definitions and referred to as such

```go

SizeMap := map[string]string{
  SMALL: "600m"
  ...
}

func getCPUSizeToCapaticy(string size){
  return SizeMap[size]
}
```

However, then the function is perhaps superfluous as you _could_ just interrogate the map directly. Perhaps you feel that is not a good API, either way, creating the maps as consts will be more efficient.

Reply:
Yeah, so I figured you'd ask about this. I don't love it, but lemme tell you what I was thinking and you can tell me what you think the right way to handle this is.

The problem is that Go doesn't support const or otherwise immutable composite types. You can't have a const map for example. So the problem I wanted to solve is I wanted these maps to be immutable; the recommendations I found for how to do this in golang is to create a method that returns a map. The map returned is still mutable obviously but at least no code can go and change the value for everyone else because the map instantiation is encapsulated in the method call.

So, that's why you see all these methods in this code that define maps and either return them or just pull something off them. I think considering Go's limits this is probably OK, but here are some ways we could change it if you like, but they're all tradeoffs:

- Just define top level maps. More efficient but more dangerous.
- Look and see if there's any kind of memoization in Go - if so we could get the safety with more efficiency
- Some other completely different way I don't know about!

Let me know what you think. Based on what I know about Go at this point I do think this is the best implementation - but I'm new to the lang so I'm very much open to a better implementation idea.

Reply:
I found multiple resources saying the same thing but this sort of article is what I based the approach on https://medium.com/bootdotdev/how-to-global-constant-maps-and-slices-in-go-30bf3eaf10db

Reply:
We could use something like this : https://github.com/kofalt/go-memoize but I don't think its a good fit. I have to imagine that the memoization code is less efficient than instantiating a map. Memoization is usually used for caching expensive methods; creating a map, though more costly than simply referencing one, should be cheaper than caching it and all the complexity that affords.

So upon further review I do think my implementation is best considering Go's limitations - but, again, I am cognizant of my own ignorance so if you know a way to solve this I'm happy to take it!

Reply:
OK, so, I think my preference here is this. Top level map, but not exported, and ensure we put this in a new module. This then means that nothing outside our module can access the map directly and that our functions to return a particular size are already inside that module, and hence we should be bale to trust what is in that small module.

Reply:
Cool - I think this actually tidies things up a bit.